### ISSUE TITLE
Android CardField getting invalid card details 

### ISSUE THREAD

**Issue by 142hitesh** [flutter-stripe/flutter_stripe#817]
**Describe the bug**
While adding card details in CardField in android getting Invalid card details even it's correct on onCardChanged. Also not getting last4 it's returning null. 

Note : Same things working in iOS. 

CardFieldInputDetails(complete: true, last4: null, expiryMonth: 5, expiryYear: 23, postalCode: null, brand: null, number: null, cvc: null, validExpiryDate: CardValidationState.Valid, validCVC: CardValidationState.Valid, validNumber: CardValidationState.Valid)

**To Reproduce**
1. Entering card number 4242424242424242 into the cardfield.
2. Entering expiry date and cvc.
3. It's getting error symbol on card icon.

**Expected behavior**
It should allow the card when all details are correct.

**Smartphone / tablet**
 - Device: Android Emulator
 - OS: Android 11
 - Package version: flutter_stripe: ^3.2.0
 - Flutter version 3.0.2

**Comment by remonh87** [comment]
I can reproduce this. @jonasbark  you have a bit more knowledge about the cardfield internals. It looks like last4 is not updated the first time but when I put the focus back on the cardnumber remove one number and add it again last4 is filled. 


I did print from android `CardChangedEvent.serializeEventData` and got: 

And got: ``` E/details   (24573): details!!! {brand=null, last4=null, expiryMonth=2, expiryYear=25, postalCode=, validNumber=Valid, validCVC=Valid, validExpiryDate=Valid}
```

Before I file an issue on react native can you double check if we did do something wrong?

**Comment by romatallinn** [comment]
Same issue. 

Is it something that came out in new releases? What is the latest workable version then?

**Comment by DavidI7** [comment]
As a temporary workaround, I was able to get the widget into a working state by setting the 'enablePostalCode' variable to true. It's not ideal but at least it's functional now.

**Comment by romatallinn** [comment]
> As a temporary workaround, I was able to get the widget into a working state by setting the 'enablePostalCode' variable to true. It's not ideal but at least it's functional now.

Thanks will know. I downgraded the plugin to 3.1.0. It seems to work there. Right now uploading it to Google Play to have the fix asap.

**Comment by jonasbark** [comment]
Opened a ticket accordingly https://github.com/stripe/stripe-react-native/issues/1041

**Comment by jonasbark** [comment]
@romatallinn @DavidI7 @142hitesh can you confirm that it works if you have enablePostalCode to false and *no* countryCode?
e.g. 
```
  CardField(
    enablePostalCode: false,
    onCardChanged: (card) {
      setState(() {
        _card = card;
      });
    },
  ),
```
seems to work for me that way.

@remonh87 let's remove sending the countryCode from the dart side, if enablePostalCode is false

**Comment by jonasbark** [comment]
https://user-images.githubusercontent.com/1151304/179233014-6af35d74-6d10-40a7-9aeb-039a0327d517.mp4

This is what I'm seeing with the example app, flutter_stripe 3.3.0, and the snippet from above. Looks fine to me.

**Comment by raulSusanj** [comment]
Hi, facing the same issue. The only workaround is what @DavidI7 posted (setting `enablePostalCode `  to `true`). @jonasbark  we use flutter_stripe 3.3.0 and the issue still exists. I am trying to figure out how to temporary avoid the enablePostalCode workaround until the package gets a fix or if I find something usefull I'll share it here.

**Comment by DavidI7** [comment]
@jonasbark Same for me, as @romatallinn already said, that potential fix you provided didn't work for me either. We're on flutter_stripe 3.3.0 too, for the record.

**Comment by jonasbark** [comment]
Can you post a video with the example app, using the code from my snippet?
I'm trying to reproduce it but can't.

**Comment by jonasbark** [comment]
A fix will be released with the next version that should arrive within the next few days

**Comment by joknjokn** [comment]
Can anyone confirm this is fixed with 4.0.0? Thanks!

**Comment by romatallinn** [comment]
> Can anyone confirm this is fixed with 4.0.0? Thanks!

Well, just wanted to drop a message that for my production app, it become even worse. Even `enablePostalCode`now doesn’t help. 😅

But it’s just me… I didn’t have a chance to test it on basic setup, coz right now downgrading the plugin to 3.3.0 and deploying to Google Play, so to have at least some working version of it. 

Will try to test it on basic app later today, but if anyone is experiencing same issue or can test it earlier, please do it.

**Comment by joknjokn** [comment]
@romatallinn Thanks for the reply! We're aiming for an urgent hotfix.

Can you describe how 4.0.0 is worse?

**Comment by romatallinn** [comment]
> @romatallinn Thanks for the reply! We're aiming for an urgent hotfix.
> 
> Can you describe how 4.0.0 is worse?

The card field accepts the card just fine now. Previously, the icon was becoming a ❗️. But instead, it now fails on card submission. Same error, card details not completed.

worth mentioning though, we are working with setup intents. So it might be different for other strategies.

**Comment by joknjokn** [comment]
@romatallinn Thank you! Sounds like 4.0.0 is not our solution.

@jonasbark Could you prioritize this? It's a very, very problematic bug.

**Comment by romatallinn** [comment]
> Thank you! Sounds like 4.0.0 is not our solution.

Do you have an opportunity to test it now as well?

**Comment by joknjokn** [comment]
@romatallinn We could potentially do that, yes. It's a little late before the weekend, so no promises, but I'll get back to you if our QA'ers find time to test 4.0.0. But I'm currently leaning towards trying to downgrade to 3.1.0.

**Comment by romatallinn** [comment]
@joknjokn okay, don't rely on anything I am saying. I downgraded to 3.3.0 and it also doesn't work the same. Probably I am having some other issue.

**Comment by romatallinn** [comment]
Yeah, it's a completely different issue that resulted in a similar error and Android only (related to #382). So, @joknjokn, sorry for the confusion. Release 4.0.0 does seem to work.

**Comment by joknjokn** [comment]
@romatallinn Yep - our QA'er tested with 4.0.0 and it does seem to be fixed 👍

**Comment by bubbler3** [comment]
Not working on RealMe X50 Android 11 with below code and flutter stripe version 4.0.0
Still need to use enablePostCode = true

```
`class NoWebhookPaymentScreen extends StatefulWidget {
  @override
  _NoWebhookPaymentScreenState createState() => _NoWebhookPaymentScreenState();
}

class _NoWebhookPaymentScreenState extends State<NoWebhookPaymentScreen> {
  final controller = CardEditController();

  @override
  void initState() {
    controller.addListener(update);
    super.initState();
  }

  void update() => setState(() {});
  @override
  void dispose() {
    controller.removeListener(update);
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ExampleScaffold(
      title: 'Card Field',
      tags: ['No Webhook'],
      padding: EdgeInsets.symmetric(horizontal: 16),
      children: [
        CardField(
          controller: controller,
        ),
        SizedBox(height: 20),
        LoadingButton(
          text: 'Pay',
          onPressed: controller.complete ? _handlePayPress : null,
        ),
        SizedBox(height: 20),
        Divider(),
        Container(
          padding: EdgeInsets.all(8),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              OutlinedButton(
                onPressed: () => controller.focus(),
                child: Text('Focus'),
              ),
              SizedBox(width: 12),
              OutlinedButton(
                onPressed: () => controller.blur(),
                child: Text('Blur'),
              ),
              SizedBox(width: 12),
              OutlinedButton(
                onPressed: () => controller.clear(),
                child: Text('Clear'),
              ),
            ],
          ),
        ),
        Divider(),
        SizedBox(height: 20),
        ResponseCard(
          response: controller.details.toJson().toPrettyString(),
        )
      ],
    );
  }

  Future<void> _handlePayPress() async {
    if (!controller.complete) {
      return;
    }

    try {
      // 1. Gather customer billing information (ex. email)
      final billingDetails = BillingDetails(
        email: 'email@stripe.com',
        phone: '+48888000888',
        address: Address(
          city: 'Houston',
          country: 'US',
          line1: '1459  Circle Drive',
          line2: '',
          state: 'Texas',
          postalCode: '77063',
        ),
      ); // mocked data for tests

      PaymentMethodParams paymentMethodParams = PaymentMethodParams.card(
          paymentMethodData: PaymentMethodData(
            billingDetails: billingDetails,
          ),
          options: PaymentMethodOptions(
              setupFutureUsage: PaymentIntentsFutureUsage.OnSession));
      // 2. Create payment method
      print('createPaymentMethod');
      final paymentMethod =
          await Stripe.instance.createPaymentMethod(paymentMethodParams);

      // 3. call API to create PaymentIntent
      print('createPaymentIntent');
      Map<String, dynamic> paymentIntentResult = await createPaymentIntent(
          paymentMethodId: paymentMethod.id,
          customerId: context.read<Customer>().id,
          amount: 25,
          currency: context.read<PaymentData>().currency);
      if (paymentIntentResult['error'] != null) {
        // Error during creating or confirming Intent
        ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(content: Text('Error: ${paymentIntentResult['error']}')));
        return;
      }

      if (paymentIntentResult['client_secret'] != null &&
          paymentIntentResult['requires_action'] == null) {
        // Payment succedeed

        ScaffoldMessenger.of(context).showSnackBar(SnackBar(
            content:
                Text('Success!: The payment was confirmed successfully!')));
        return;
      }

      if (paymentIntentResult['client_secret'] != null &&
          paymentIntentResult['requires_action'] == true) {
        // 4. if payment requires action calling handleNextAction
        print('handleNextAction');
        final paymentIntent = await Stripe.instance
            .handleNextAction(paymentIntentResult['client_secret']);
        // todo handle error
        /*if (cardActionError) {
        Alert.alert(
        `Error code: ${cardActionError.code}`,
        cardActionError.message
        );
      } else*/

        if (paymentIntent.status == PaymentIntentsStatus.RequiresConfirmation) {
          // 5. Call API to confirm intent
          print('confirmIntent');
          await confirmIntent(paymentIntent, paymentMethodParams);
        } else {
          // Payment succedeed
          ScaffoldMessenger.of(context).showSnackBar(SnackBar(
              content: Text('Error: ${paymentIntentResult['error']}')));
        }
      }
    } catch (e) {
      ScaffoldMessenger.of(context)
          .showSnackBar(SnackBar(content: Text('Error: $e')));
      rethrow;
    }
  }

  Future<void> confirmIntent(PaymentIntent paymentIntent,
      PaymentMethodParams paymentMethodParams) async {
    final paymentIntent2 = await Stripe.instance.confirmPayment(
      paymentIntent.clientSecret,
      PaymentMethodParams.card(
        paymentMethodData: PaymentMethodData(
          billingDetails: BillingDetails(
            email: 'email@stripe.com',
            phone: '+48888000888',
            address: Address(
              city: 'Houston',
              country: 'US',
              line1: '1459  Circle Drive',
              line2: '',
              state: 'Texas',
              postalCode: '77063',
            ),
          ),
        ),
      ),
    );
  }
}
`
```

**Comment by davemg3** [comment]
Not working on RealMe X50 Android 11 with below code and flutter stripe version 4.0.0
Still need to use enablePostCode = true to make the card details valid.


```
`class NoWebhookPaymentScreen extends StatefulWidget {
  @override
  _NoWebhookPaymentScreenState createState() => _NoWebhookPaymentScreenState();
}

class _NoWebhookPaymentScreenState extends State<NoWebhookPaymentScreen> {
  final controller = CardEditController();

  @override
  void initState() {
    controller.addListener(update);
    super.initState();
  }

  void update() => setState(() {});
  @override
  void dispose() {
    controller.removeListener(update);
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ExampleScaffold(
      title: 'Card Field',
      tags: ['No Webhook'],
      padding: EdgeInsets.symmetric(horizontal: 16),
      children: [
        CardField(
          controller: controller,
        ),
        SizedBox(height: 20),
        LoadingButton(
          text: 'Pay',
          onPressed: controller.complete ? _handlePayPress : null,
        ),
        SizedBox(height: 20),
        Divider(),
        Container(
          padding: EdgeInsets.all(8),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              OutlinedButton(
                onPressed: () => controller.focus(),
                child: Text('Focus'),
              ),
              SizedBox(width: 12),
              OutlinedButton(
                onPressed: () => controller.blur(),
                child: Text('Blur'),
              ),
              SizedBox(width: 12),
              OutlinedButton(
                onPressed: () => controller.clear(),
                child: Text('Clear'),
              ),
            ],
          ),
        ),
        Divider(),
        SizedBox(height: 20),
        ResponseCard(
          response: controller.details.toJson().toPrettyString(),
        )
      ],
    );
  }

  Future<void> _handlePayPress() async {
    if (!controller.complete) {
      return;
    }

    try {
      // 1. Gather customer billing information (ex. email)
      final billingDetails = BillingDetails(
        email: 'email@stripe.com',
        phone: '+48888000888',
        address: Address(
          city: 'Houston',
          country: 'US',
          line1: '1459  Circle Drive',
          line2: '',
          state: 'Texas',
          postalCode: '77063',
        ),
      ); // mocked data for tests

      PaymentMethodParams paymentMethodParams = PaymentMethodParams.card(
          paymentMethodData: PaymentMethodData(
            billingDetails: billingDetails,
          ),
          options: PaymentMethodOptions(
              setupFutureUsage: PaymentIntentsFutureUsage.OnSession));
      // 2. Create payment method
      print('createPaymentMethod');
      final paymentMethod =
          await Stripe.instance.createPaymentMethod(paymentMethodParams);

      // 3. call API to create PaymentIntent
      print('createPaymentIntent');
      Map<String, dynamic> paymentIntentResult = await createPaymentIntent(
          paymentMethodId: paymentMethod.id,
          customerId: context.read<Customer>().id,
          amount: 25,
          currency: context.read<PaymentData>().currency);
      if (paymentIntentResult['error'] != null) {
        // Error during creating or confirming Intent
        ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(content: Text('Error: ${paymentIntentResult['error']}')));
        return;
      }

      if (paymentIntentResult['client_secret'] != null &&
          paymentIntentResult['requires_action'] == null) {
        // Payment succedeed

        ScaffoldMessenger.of(context).showSnackBar(SnackBar(
            content:
                Text('Success!: The payment was confirmed successfully!')));
        return;
      }

      if (paymentIntentResult['client_secret'] != null &&
          paymentIntentResult['requires_action'] == true) {
        // 4. if payment requires action calling handleNextAction
        print('handleNextAction');
        final paymentIntent = await Stripe.instance
            .handleNextAction(paymentIntentResult['client_secret']);
        // todo handle error
        /*if (cardActionError) {
        Alert.alert(
        `Error code: ${cardActionError.code}`,
        cardActionError.message
        );
      } else*/

        if (paymentIntent.status == PaymentIntentsStatus.RequiresConfirmation) {
          // 5. Call API to confirm intent
          print('confirmIntent');
          await confirmIntent(paymentIntent, paymentMethodParams);
        } else {
          // Payment succedeed
          ScaffoldMessenger.of(context).showSnackBar(SnackBar(
              content: Text('Error: ${paymentIntentResult['error']}')));
        }
      }
    } catch (e) {
      ScaffoldMessenger.of(context)
          .showSnackBar(SnackBar(content: Text('Error: $e')));
      rethrow;
    }
  }

  Future<void> confirmIntent(PaymentIntent paymentIntent,
      PaymentMethodParams paymentMethodParams) async {
    final paymentIntent2 = await Stripe.instance.confirmPayment(
      paymentIntent.clientSecret,
      PaymentMethodParams.card(
        paymentMethodData: PaymentMethodData(
          billingDetails: BillingDetails(
            email: 'email@stripe.com',
            phone: '+48888000888',
            address: Address(
              city: 'Houston',
              country: 'US',
              line1: '1459  Circle Drive',
              line2: '',
              state: 'Texas',
              postalCode: '77063',
            ),
          ),
        ),
      ),
    );
  }
}
```

**Comment by sleewok** [comment]
> @romatallinn Yep - our QA'er tested with 4.0.0 and it does seem to be fixed 👍

Correct, it is NOT fixed in 4.0.0. I am using the CardFormField. The only way it works is if I enable postalCode.

**Comment by jonasbark** [comment]
@sleewok I can't confirm that. Please provide an example project where I can see that it does not work. Looks like you still use the old version, so better double-check.

**Comment by sleewok** [comment]
> @sleewok I can't confirm that. Please provide an example project where I can see that it does not work. Looks like you still use the old version, so better double-check.

Definitely running the latest version:
C:\flutter\.pub-cache\hosted\pub.dartlang.org\flutter_stripe-4.0.0\lib\flutter_stripe.dart

**I think I need to clarify something.** My error may be different. 
First, No error is showing on the widget GUI for the CardFormField. 

However, the form is never marked as completed UNLESS a postalCode is entered along with the other information (CC, CVC, EXP). This is the case even if I have set postalCode = false.

This is what the formDetails are with the postalCode = false and the CC, CVC, EXP all entered correctly: 
`CardFieldInputDetails(complete: false, last4: null, expiryMonth: null, expiryYear: null, postalCode: null, brand: null, number: null, cvc: null, validExpiryDate: CardValidationState.Unknown, validCVC: CardValidationState.Unknown, validNumber: CardValidationState.Unknown)`

If I set postalCode = true, and enter the postal code the result changes and is marked as complete as soon as I complete all the fields. It will toggle complete if I delete the CVC, etc, just as expected:
`CardFieldInputDetails(complete: true, last4: 4242, expiryMonth: 4, expiryYear: 2024, postalCode: 24982, brand: Visa, number: null, cvc: null, validExpiryDate: CardValidationState.Unknown, validCVC: CardValidationState.Unknown, validNumber: CardValidationState.Unknown)`

I'd rather not have the user enter the postal in the form, since I collect address information prior to the CC entry.
